progen addition

0xc0170 opened the pull request:
Adding progen to Exporter class. This is just a boilerplate for the future addition/replacement for tool exporters. I sent earlier bigger PR to replace uvision/iar. This should be easier to integrate, as it should not change anything.

This enables anyone to add new tools or replace our current tools by progen. To see how to, look at my other branch, uvision4 integration: https://github.com/0xc0170/mbed/blob/62950c9fc4b75170971a096b0f548b65695f139c/workspace_tools/export/uvision4.py

@sg- @bridadan @screamerbg

bridadan commented on workspace_tools/export/exporters.py:
```diff
@@ -12,12 +12,99 @@
     DOT_IN_RELATIVE_PATH = False
 
+    # mapping mbed to progen targets
+    PROGEN_TARGETS = {
```
Have you thought about moving this definition into the target's definition in targets.py?

It could either be a separate attribute, like `self.exporters_supported = { 'progen': 'mbed-lpc1768' }` or `self.exporters_supported = { 'progen': True }` (to use the default name) for the LPC1768. Not too sure how it should look, but I'm worried about maintaining a separate list of target mappings that people can miss when they're adding new platforms.

0xc0170 replied:
We can add a new member to a target:

```
MY_TARGET():
  self.progen_target = 'mbed-lpc1768'
```

Then in Exporters, we would try to get that one, otherwise use the target name as it is now ?

bridadan replied:
Makes sense to me, although is it possible that progen will not support a given target? If so, we should have some way of indicating that.

Or more ideally, each target is not supported by default and we specify which ones are.

sg- commented:
@bridadan That is a great idea!

0xc0170 commented:
Rebased, added progen_target to Targets. It is used only when progen is active for exporters (currently none). I have tested it with uvision locally, I'll send a new PR with uvision progen exporter later.

bridadan commented:
test this please

HOST_OSES=ALL
BUILD_TOOLCHAINS=ALL
TARGETS=ALL_CONNECTED

bridadan commented:
Changes look good to me. Even though the build bot says failure, those are existing issues and your changes did not cause new regressions.

0xc0170 commented:
@bridadan Thanks ! I'll wait one more day for review, and will merge this week
